posts/inside-rust/test-infra-jan-feb-2025.md

```diff
@@ -0,0 +1,204 @@
+
+### `rustc`-based (`ToolRustc`) tools have unified staging handling
+
+Tools that wants to use a locally built `rustc` previously inconsistently implemented their own staging logic in their tool and test steps. This caused a lot of confusion as different `ToolRustc` tools (and their tests) handled the staging differently; some had unnecessarily builds while others were seemingly "off by one stage". There were hacks in various places to "chop off" or "increment" stages separately. To make this situation more maintainable, <https://github.com/rust-lang/rust/pull/137215> unifies `ToolRustc` tool staging logic.
+
+Notably, `./x test` without args and `./x test src/tools/{cargo,clippy}`, where possible, now default to stage 2. Previously, `./x test src/tools/{cargo,clippy}` without explicit test stage configuration corresponded to `--stage 1` but they actually required building stage 2 rustc *anyway*. Bootstrap will now warn if you try to specify a test stage < 2 when testing these two tools (that they don't necessarily work against stage 1 rustc is an pre-existing issue).
+
+Consult the new [Writing tools in Bootstrap](https://github.com/rust-lang/rust/blob/master/src/doc/rustc-dev-guide/src/building/bootstrapping/writing-tools-in-bootstrap.md) chapter for further clarification on picking a correct bootstrap tool mode.
+
+### `run-make-support` and `rmake.rs` is now fixed to be built with stage 0 compiler
+
+See <https://github.com/rust-lang/rust/pull/137373> and <https://github.com/rust-lang/rust/pull/137537>.
+
+Previously, `run-make-support` and `rmake.rs` was mistakenly built with top-stage compiler, but this was wrong. `run-make-support` and `rmake.rs` should be built with the *stage 0* compiler (they are test *infra* and needs to be reliable regardless of the possibly broken stage > 0 compiler under test). This caused a few `rmake.rs` tests to accidentally be using unstable features in the test recipes themselves, which will cause issues for beta/stable backports/bumps, and will also cause issues for out-of-tree codegen backends like `rustc_codegen_cranelift` that needs to run `run-make` tests at stage 0.
+
+The docs are also updated to explicitly clarify that `run-make-support` and `rmake.rs` *may not use unstable features*.
+
+### `core` unit tests are now located in a separate `coretests` package
+
+Having std tests in the same package as a std crate has issues such as
+
+> \[causing\] the test to have a dependency on a locally built standard library crate, while also indirectly depending on it through `libtest`
+
+<https://github.com/rust-lang/rust/pull/135937> moves `core` tests into a separate package that *does not* depend on `core` to prevent the duplicate crates problem, even when compiler flags don't match between sysroot build and test build.
+
+Other parts of std (such as `alloc`) still has this problem. This is part of an on-going effort to make std tests more robust and more buildable by custom codegen backends.
```

Comment:
alloc was ported in https://github.com/rust-lang/rust/pull/136642. std still has this issue though.

Reply:
Ah for some reason I thought that was still open. I'll update it.

Reply:
Fixed, updated this to reflect both `coretests` and `alloctests` are now separate packages.